### DISCUSSION TITLE
Why does code coverage in a monorepo require a root vitest.config.ts?

### DISCUSSION THREAD

**Discussion by kieranm** [vitest-dev/vscode#454] [Q&A] [answered]
Hey all,

I'm setting up code coverage in a monorepo. Previously I had...

- A `vitest.workspace.ts` to configure package locations
- `vitest.config.ts` in each package.

However I found that with this setup, no code coverage was being output.

When I added the following extra `vitest.config.ts` file to the root of the workspace I started getting coverage output as expected.

```
import { defineConfig } from "vitest/config"

export default defineConfig({
  test: {
    globals: true,
    environment: "node",
    coverage: {
      enabled: true,
      provider: "v8",
    },
  },
})

```

It's not clear to me why this is needed - why doesn't the extension pick up the coverage config from the individual packages? Have I missed something? It's also not clear to me whether I should be configuring coverage in this root file or in the individual workspace config files.

Thanks
Kieran

**Comment by sheremet-va** [comment, marked as answer]
`coverage` field never worked with workspace project's configs:

- [vitest.dev/guide/workspace.html#configuration](https://vitest.dev/guide/workspace.html#configuration)
- [vitest.dev/config/#coverage](https://vitest.dev/config/#coverage)

Even with Vitest CLI, you will not get the coverage if you run it from the root. Of course, if you run tests from the project directory, Vitest will pick up its config and won't use workspaces at all - then it will print the coverage, but it will be scoped only to the project itself.

This limitation exists because coverage is reported for _every_ workspace project - you can't reconfigure it midthrough. That's the whole point of the workspace.

**Reply by kieranm** [reply]
So just so I understand, `coverage` is ignored if it appears in a workspace project config, and the config for coverage is defined globally in the `vitest.config.ts` in the workspace root?

If pressing the ▶️ button in VSCode to run the tests with coverage, are they run from the root? Does the project level config get picked up?

Thanks

**Reply by sheremet-va** [reply]
It depends on your [VSCode workspace](https://code.visualstudio.com/docs/editor/workspaces). By default, it will probably use the root `vitest.workspace.ts` and `vitest.config.ts` files (when just a single root folder is opened in VSCode)
